borsh/src/de/mod.rs

```diff
@@ -533,17 +536,26 @@ const _: () = {
         $(
             impl<T> BorshDeserialize for [T; $len]
             where
-                T: BorshDeserialize + Default + Copy
+                T: BorshDeserialize
             {
                 #[inline]
                 fn deserialize(buf: &mut &[u8]) -> Result<Self> {
-                    let mut result = [T::default(); $len];
+
+                    let mut result: [MaybeUninit<T>; $len] = unsafe { MaybeUninit::uninit().assume_init() };
+
                     if !T::copy_from_bytes(buf, &mut result)? {
-                        for i in 0..$len {
-                            result[i] = T::deserialize(buf)?;
+                        for elem in &mut result {
+                            elem.write(T::deserialize(buf)?);
```

Comment:
This will leak `T` if deserialize reteruns an error. This feels like a big bug: if the server expects `[String; 4]`, I can OOM it by sending shorter-than-necessary buffer. 

`array::from_fn` is sadly unstable as of yet, so we'd have to vendor that it (and the impl is a bit non-trivial. )

Reply:
I think you meant to say `array::try_from_fn`? Yeah, it seems to have its roots pretty deep. Unsure if it's worth.

Very valid on the leak, would it be too janky to try deserializing, and on failure loop through and drop all T that have been placed? Just trying to avoid vendoring a ton of code

Reply:
Yeah, that I think is broadly reasonable. I'd say it needs two chages: 

* put the logic into some sort of a drop gurard, such that *panic* in deseralize also triggers dropping of inited stuff
* the pedantic logic would also try to drop the rest of the elements even if dropping one of them panicked. I *think* you get get that for free by calling `ptr::drop_in_place(&mut [T])`

Reply:
Did a quick workaround [`692cd8a` (#104)](https://github.com/near/borsh-rs/pull/104/commits/692cd8a1237b91022c6eaf02c1c4b829a9c2beef) but going to clean up tomorrow. I'll open this PR back up for review when it's ready :)

Reply:
Marking PR as ready now. I'm still not 100% confident in the current impl, but it's hard for me to come up with an argument against it.

The main change, aside from switching to use `drop_in_place`, is to change wrapping the arrays in a `MaybeUninit` and then assuming init to just cast as a pointer to `[T; N]` before copying (in the case of initializing) or to `*mut [T]` for `drop_in_place` (https://github.com/near/borsh-rs/pull/104/commits/d26078e864f5a3aaf78404a374fd65bd638b170a). I can't pinpoint what the intermediate type is and why it's safe to cast the pointer like this, and if this is actually safe to make the assumption that the memory layout of `[T]` will always be the same as `[MaybeUninit<T>]`. Do you have an argument against or a suggestion here?

Reply:
>I can't pinpoint what the intermediate type is


Not sure I understand the Q, but maybe this: `init_range as *mut [MaybeUninit<T>] as *mut [T]`?

>why it's safe to cast the pointer like this

Pointer casting is always safe, only dereferencing the result can be UB

>that the memory layout of [T] will always be the same as [MaybeUninit<T>]

Yeah, I think [T] == [MaybeUnint<T>] is guaranteed, though not-necessary spelled out. This is in contrast to Option<T> vs Option<MaybeUninit<T>> which doesn't hold due to niches.

Reply:
> Not sure I understand the Q, but maybe this: `init_range as *mut [MaybeUninit<T>] as *mut [T]`?

Okay, I might be going crazy; I swear I tried exactly that before, and it gave a compiler error. Can't remember what toolchain I was using at the time, though. Nevermind :D 

> Yeah, I think [T] == [MaybeUnint] is guaranteed

Thanks, I should have been more thorough through the docs and I would have seen https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#layout-1 also